driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java

```diff
@@ -41,25 +40,40 @@
 import java.util.function.Supplier;
 
 import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
 
 public final class MongoCryptHelper {
 
-    public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<String, Object>> kmsProviders) {
-        return createMongoCryptOptions(kmsProviders, null, null, false);
+    public static MongoCryptOptions createMongoCryptOptions(final ClientEncryptionSettings settings) {
+        return createMongoCryptOptions(settings.getKmsProviders(), false, emptyList(), emptyMap(), null, null);
     }
 
-    public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<String, Object>> kmsProviders,
+    public static MongoCryptOptions createMongoCryptOptions(final AutoEncryptionSettings settings) {
+        return createMongoCryptOptions(
+                settings.getKmsProviders(),
+                settings.isBypassQueryAnalysis(),
+                settings.getSearchPaths(),
+                settings.getExtraOptions(),
+                settings.getSchemaMap(),
+                settings.getEncryptedFieldsMap());
+    }
+
+    private static MongoCryptOptions createMongoCryptOptions(
+            final Map<String, Map<String, Object>> kmsProviders,
+            final boolean bypassQueryAnalysis,
+            final List<String> searchPaths,
+            @Nullable final Map<String, Object> extraOptions,
             @Nullable final Map<String, BsonDocument> localSchemaMap,
-            @Nullable final Map<String, BsonDocument> encryptedFieldsMap,
-            final boolean bypassQueryAnalysis) {
+            @Nullable final Map<String, BsonDocument> encryptedFieldsMap) {
         MongoCryptOptions.Builder mongoCryptOptionsBuilder = MongoCryptOptions.builder();
-
-        BsonDocument bsonKmsProviders = getKmsProvidersAsBsonDocument(kmsProviders);
-        mongoCryptOptionsBuilder.kmsProviderOptions(bsonKmsProviders);
-        mongoCryptOptionsBuilder.needsKmsCredentialsStateEnabled(true);
+        mongoCryptOptionsBuilder.kmsProviderOptions(getKmsProvidersAsBsonDocument(kmsProviders));
+        mongoCryptOptionsBuilder.bypassQueryAnalysis(bypassQueryAnalysis);
+        mongoCryptOptionsBuilder.searchPaths(searchPaths);
```

Comment:
Are these two lines the only functional changes in this class?  It's hard to tell with all the refactoring mixed in.

Reply:
Functionally these are the two new options passed to mongoCryptOptionsBuilder.  

The refactorings were mainly to support this:

`toBsonDocument` was extracted to handle multiple wrapping of maps - which then could be used in fetchCredentials.

The other refactorting is simplifying the public `createMongoCryptOptions` options by passing in the actual settings and configuring the defaults here - and making a private createMongoCryptOptions method with all the arguments that can be used.

The only drive by refactoring was using closures in the creation of MongoClientSettings.

Reply:
OK, I see your point.